_posts/2023-04-13-floating-point-error-analysis-with-chef-fp.md

```diff
@@ -0,0 +1,116 @@
+the presented framework.
+
+CHEF-FP automates error estimation by injecting error calculation code into
+the generated derivatives, enhancing analysis time and reducing memory usage.
+The framework's ability to provide Floating-Point error analysis and generate
+optimized code contributes to significant speed-ups in the analysis time.
+
+### Modeling Floating-Point Errors
+
+The error model describes a metric that can be used in the error estimation
+analysis. This metric is applied to all assignment operations in the
+considered function, and the results from its evaluation are accumulated into
+the total floating point error of the left-hand side of the assignment. This
+error model is sufficient for most smaller cases and can produce loose upper
+bounds of the maximum permissible FP error in programs.
+
+### Automatic Differentiation (AD) Basics
+
+With increasingly better language support, AD is becoming a much more
+attractive tool for researchers. AD takes as input any program code that has
+meaningful differentiable properties and then it produces a new code that is
+augmented with pushforward (forward mode AD) or pullback (reverse mode AD)
+operators. 
+
+The Reverse Mode AD in particular provides an efficient way to compute the
+function’s gradient with relative time complexity,  which is independent of
+the size of the input. Among AD techniques, the Source Transformation approach
+has better performance, since it does as much as possible at compile time to
+create the derivative only once. 
+
+### AD-Based Floating-Point Error Estimation Using CHEF-FP
+
+An important part of dealing with floating-point applications with high
+precision requirements is identifying the sensitive (more error-prone) areas
+to implement suitable mitigation strategies. This is where AD-based
+sensitivity analysis can be very useful. It can find specific variables or
+regions of code with a high contribution to the overall FP error in the
+application. 
+
+The CHEF-FP framework requires less manual integration work and comes packaged
+with Clad, taking away the tedious task of setting up long tool chains. The
+tool’s proximity to the compiler (and the fact that the floating-point error
+annotations are built into the code’s derivatives) allows for powerful
+compiler optimizations that provide significant improvements in the analysis
+time. Another upside of using a source-level AD tool like Clad as the backend
+is that it provides important source insights (variable names, IDs, source
+location, etc.). Clad can also identify special constructs (such as loops,
+lambdas, functions, if statements, etc.) and fine-tune the error code
+generation accordingly. 
+
+### Implementation
+
+CHEF-FP's implementation involves registering an API in Clad for calculating
+the floating-point errors in desired functions using the default error
+estimation models. 
+
+While CHEF-FP provides an API that is useful for building simple expressions,
```

Comment:
maybe we can rephrase in a positive way? Eg, CHEF-FP provides two APIs...

Reply:
Updated Text:

_CHEF_FP not only serves as a proof-of-concept, but it also provides APIs for building common expressions. For more complex expressions, custom calls to external functions can be built as a valid [custom error model](https://github.com/vgvassilev/clad/blob/v1.1/demos/ErrorEstimation/CustomModel/README.md), as long as the function has a compatible return type for the variable being assigned the error. This means that users can define their error models as regular C++ functions, enabling the implementation of more computationally complex models._